0000-settlement-engine/0000-settlement-engine.md

```diff
@@ -0,0 +1,415 @@
+
+##### FULFILL
+
+- `fulfillment`: `0000000000000000000000000000000000000000000000000000000000000000`
+- `data`: _Response message from recipient settlement engine_
+
+##### REJECT
+
+- `code`: _Determined by connector from HTTP status of forwarded request_
+- `triggeredBy`: `peer.settle`
+- `message`: _Determined by connector_
+- `data`: _Response message from recipient settlement engine, or empty if antecedent failure_
+
+### Motivation
+
+Settlement engines supercede the [Ledger Plugin Interface (LPIv2)](../deprecated/0024-ledger-plugin-interface-2/0024-ledger-plugin-interface-2.md), an earlier abstraction for settlement integrations with Interledger. This new model addresses these issues:
+
+1. Multi-account plugins required logic for handling ILP packets, increasing implementation complexity.
+2. Plugins bundled settlement and bilateral communication functionality together, limiting composability.
+3. JavaScript plugins limited interoperability with non-JavaScript connector implementations.
+4. Plugins operated in the same process as the connector, which limited scaling the two services independently.
+
+## Specification
+
+### Accounts and identifiers
+
+Each account MUST be identified by a unique, [URL-safe](https://tools.ietf.org/html/rfc3986#section-2.3) string.
+
+The settlement engine MUST be responsible for correlating an account identifier to the peer's identity on the shared ledger or settlement system, if required. For separation of concerns between clearing and settlement, the accounting system is NOT RECOMMENDED to have knowledge of the peer's identity on the shared settlement system.
+
+### Units and quantities
+
+Asset amounts may be represented using different denominations. For example, one U.S. dollar may be represented as \$1 or 100 cents, each of which is equivalent in value. Likewise, one Bitcoin may be represented as 1 BTC or 100,000,000 satoshis.
+
+A **standard unit** is the typical unit of value for a particular asset, such as \$1 in the case of U.S. dollars, or 1 BTC in the case of Bitcoin.
+
+A **fractional unit** represents some unit smaller than the standard unit, but with greater precision. Examples of fractional monetary units include one cent (\$0.01 USD), or 1 satoshi (0.00000001 BTC).
+
+An **asset scale** is the difference in orders of magnitude between a standard unit and a corresponding fractional unit. More formally, the asset scale is a non-negative integer (0, 1, 2, …) such that one standard unit equals `10^(-scale)` of a corresponding fractional unit. If the fractional unit equals the standard unit, then the asset scale is 0.
+
+For example, one cent represents an asset scale of 2 in the case of USD, whereas 1 satoshi represents an asset scale of 8 in the case of Bitcoin.
+
+#### Selecting scales
+
+Settlement engines MUST use the denomination of its settlements, which is typically the smallest denomination of the asset on the settlement system, as the scale to perform or fulfill requests.
```

Comment:
I think we should strike this line. It's sort of a confusing sentence to grok with outsider glasses on (e.g., "Settlement engines MUST use the denomination of its settlements...", while true, sounds like a sylogism). 

Also, technically speaking, if a SE wanted to do something internally like this and could figure out a way to do it internally that didn't violate the externally-facing contract of this RFC, then so be it. 

In other words, I think line 203 is enough here.

Reply:
Agree this version is confusing... am open to nixing it. Before we do, is this version any better?

"Settlement engines MUST perform or fulfill requests with amounts denoted in the same denomination as its settlements, which are typically the smallest denomination of the asset on the settlement system." 

> Also, technically speaking, if a SE wanted to do something internally like this and could figure out a way to do it internally that didn't violate the externally-facing contract of this RFC, then so be it.

While this is true, my concern is that SE implementers might try to support greater precision than can actually be settled within the SE itself -- which should be expressly disallowed, since that's better served as a feature of an accounting system.

Reply:
> my concern is that SE implementers might try to support greater precision than can actually be settled within the SE itself 

This seems unlikely, or at least pretty difficult. If I'm understanding the idea here, a hypothetical XRPL SE would be configured to accept settlements in milli-drops, but because this can't actually occur on the _actual_ XRP Ledger, the SE would need to buffer these payments until they became large enough to equal one drop.

I agree this is probably not advisable, but who cares if a Settlement Engine tries to do this? Maybe a use-case can make it happen in a tolerable manner.

> should be expressly disallowed
I don't think this should be expressly disallowed, right? Seems like we should append this to line 203:

"Additionally, it is RECOMMENDED that settlement engines do not advertise denominations that are smaller than the smallest value supported by their underlying ledger system."

Reply:
I think this is resolved now in the latest update.